SYCL/ESIMD/api/functional/functions/functions_atomic_update_inc_dec.cpp

```diff
@@ -0,0 +1,266 @@
+    for (size_t i = 0; i < NumElemsToChange; ++i) {
+      values_sum += values_before_update[i];
+    }
+
+    return values_sum;
+  }
+};
+
+// Provides expected value for updated memory locations.
+template <esimd::atomic_op Operator, typename T>
+T get_expected_value(T base_value, int number_of_iterations) {
+  if constexpr (Operator == esimd::atomic_op::dec) {
+    return base_value - number_of_iterations;
+  } else if constexpr (Operator == esimd::atomic_op::inc) {
+    return base_value + number_of_iterations;
+  } else {
+    static_assert(Operator != Operator, "Unexpected  operator type.");
+  }
+}
+
+// The main test routine.
+// Using functor class to be able to iterate over the pre-defined data types.
+template <typename TestCaseT, typename DataT, typename NumElemsT,
+          typename NDRangeDimT, typename NumElemsToChangeT,
+          typename ChangeElemFilterT, typename ChosenOperatorT,
+          typename AlgorithmToChangeT>
+class run_test {
+  static constexpr int NumElems = NumElemsT::value;
+  static constexpr int NDRangeDim = NDRangeDimT::value;
+  static constexpr int NumElemsToChange = NumElemsToChangeT::value;
+  static constexpr atomic_op ChosenOperator = ChosenOperatorT::value;
+  static constexpr functions::offset_generation AlgorithmToChange =
+      AlgorithmToChangeT::value;
+  using TestDescriptionT = functions::TestDescription<NumElems, TestCaseT>;
+
+public:
+  bool operator()(sycl::queue &queue, const std::string &data_type,
+                  const std::string &operator_type) {
+    static_assert(NDRangeDim <= 3 && NDRangeDim > 0);
+    static_assert(NumElemsToChange <= NumElems);
+    bool passed = true;
+    constexpr DataT base_value = 100;
+    constexpr DataT result_fill_value = 5;
+    shared_allocator<DataT> allocator(queue);
+    shared_vector<DataT> shared_init_values(NumElems, allocator);
+    std::fill(shared_init_values.begin(), shared_init_values.end(), base_value);
+
+    shared_vector<size_t> shared_offsets(NumElemsToChange,
+                                         shared_allocator<size_t>(queue));
+    functions::fill_offsets<NumElemsToChange, AlgorithmToChange>(
+        shared_offsets);
+    shared_vector<size_t> local_offsets = shared_offsets;
+    constexpr int NumberIterations = 16;
+    shared_vector<DataT> shared_result(NumberIterations, result_fill_value,
+                                       allocator);
+    sycl::nd_range<NDRangeDim> nd_range =
+        get_sycl_nd_range<NDRangeDim>(NumberIterations);
+
+    queue.submit([&](sycl::handler &cgh) {
+      DataT *shared_init_values_ptr = shared_init_values.data();
+      DataT *shared_result_ptr = shared_result.data();
+      size_t *const offsets_ptr = shared_offsets.data();
+
+      cgh.parallel_for<
+          Kernel<DataT, NumElems, TestCaseT, ChosenOperatorT, NDRangeDimT,
+                 NumElemsToChangeT, ChangeElemFilterT, AlgorithmToChangeT>>(
+          nd_range, [=](sycl::nd_item<1> nd_item) SYCL_ESIMD_KERNEL {
```

Comment:
The test seems too weak. 16 threads each doing an atomic update is likely to succeed even w/o atomic access. There should be many more threads, ideally, more than there are hardware threads, and each thread should do atomic update in a loop for some time to achieve actual contention.

Reply:
To think of, we also might use a barrier to sync the start of work-items execution.

Reply:
@kbobrovs Is it ok if we will obtain max number of threads by using `sycl::device::get_info<info::device::max_compute_units>()`?

> There should be many more threads, ideally, more than there are hardware threads


Also, number of threads should be less than `sycl::device::get_info<info::device::max_work_group_size>()`, I am right?

The following device information descriptors:
- info::device::max_compute_units
- info::device::max_work_group_size

Described in [SYCL-2020 #4.6.4.2. Device information descriptors](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#_device_information_descriptors)

Reply:
> Also, number of threads should be less than `sycl::device::get_info<info::device::max_work_group_size>()`, I am right?

`max_work_group_size()` does not reflect the number of h/w threads. It should be a function of `max_compute_units>()` (not sure about the exact formula). But NDRange can definitely exceed `max_hardware_threads * sub_group_size` . S/W threads will be mapped to hardware in "rounds". That is my suggestion.

> To think of, we also might use a barrier to sync the start of work-items execution.

It is always a good idea to reduce the number of features used in a test to make it more focused. I think it will be strong enough w/o barriers if there is a loop in each thread. Besides, we only have group barriers, no machine-wide ones.